server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -104,39 +106,26 @@ public void start() throws IOException
   {
     serverAnnouncer.announce();
 
+    fireChiefExecutor = Execs.multiThreaded(fireDepartments.size(), "chief-%d");
+
     for (final FireDepartment fireDepartment : fireDepartments) {
       final DataSchema schema = fireDepartment.getDataSchema();
 
       final FireChief chief = new FireChief(fireDepartment, conglomerate);
-      Map<Integer, FireChief> partitionChiefs = chiefs.get(schema.getDataSource());
-      if (partitionChiefs == null) {
-        partitionChiefs = new HashMap<>();
-        chiefs.put(schema.getDataSource(), partitionChiefs);
-      }
-      partitionChiefs.put(fireDepartment.getTuningConfig().getShardSpec().getPartitionNum(), chief);
-
-      chief.setName(
-          String.format(
-              "chief-%s[%s]",
-              schema.getDataSource(),
-              fireDepartment.getTuningConfig().getShardSpec().getPartitionNum()
-          )
-      );
-      chief.setDaemon(true);
-      chief.start();
+      chiefs.computeIfAbsent(schema.getDataSource(), k -> new HashMap<>())
+            .put(fireDepartment.getTuningConfig().getShardSpec().getPartitionNum(), chief);
+
+      fireChiefExecutor.submit(chief);
     }
   }
 
   @LifecycleStop
   public void stop()
   {
-    for (Map<Integer, FireChief> chiefs : this.chiefs.values()) {
-      for (FireChief chief : chiefs.values()) {
-        CloseQuietly.close(chief);
-      }
-    }
-
     serverAnnouncer.unannounce();
+    if (fireChiefExecutor != null) {
+      fireChiefExecutor.shutdownNow();
```

Comment:
Previously, fireChiefs are closed here which means their firehoses are closed as well. This is effectively same with interrupting fireChief threads because iterating on closed firehoses will throw an exception.